File: backend/routes/logs.js
const express = require("express");
const router = express.Router();
const { isAuthenticated } = require("./auth");
const logsService = require("../services/logs");

// Get available log sources
router.get("/sources", isAuthenticated, async (req, res) => {
    try {
        const sources = await logsService.getLogSources();
        res.json({ success: true, sources });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Get targets for a source
router.get("/sources/:sourceId/targets", isAuthenticated, async (req, res) => {
    try {
        const targets = await logsService.getLogTargets(req.params.sourceId);
        res.json({ success: true, targets });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

// Get logs from source
router.get("/sources/:sourceId", isAuthenticated, async (req, res) => {
    try {
        const { sourceId } = req.params;
        const { target, lines = 100, search } = req.query;

        let logs = await logsService.getLogsFromSource(sourceId, target, parseInt(lines));

        if (search) {
            logs = logsService.searchInLogs(logs, search);
        }

        res.json({ success: true, logs });
    } catch (error) {
        res.status(500).json({ success: false, error: error.message });
    }
});

module.exports = router;

File: be/internal/handlers/logs.go
package handlers

import (
	"context"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	"github.com/nawfdev/home-panel/internal/httpx"
	logsvc "github.com/nawfdev/home-panel/internal/logs"
)

type logsService interface {
	Sources(ctx context.Context) []logsvc.Source
	Targets(ctx context.Context, sourceID string) []logsvc.Target
	Logs(ctx context.Context, sourceID, target string, lines int) string
}

// Logs ports backend/routes/logs.js.
type Logs struct {
	Svc logsService
}

func (l *Logs) Sources(w http.ResponseWriter, r *http.Request) {
	httpx.JSON(w, http.StatusOK, map[string]any{"success": true, "sources": l.Svc.Sources(r.Context())})
}

func (l *Logs) Targets(w http.ResponseWriter, r *http.Request) {
	targets := l.Svc.Targets(r.Context(), chi.URLParam(r, "sourceId"))
	httpx.JSON(w, http.StatusOK, map[string]any{"success": true, "targets": targets})
}

func (l *Logs) Source(w http.ResponseWriter, r *http.Request) {
	lines := 100
	if raw := r.URL.Query().Get("lines"); raw != "" {
		if parsed, err := strconv.Atoi(raw); err == nil {
			lines = parsed
		}
	}

	out := l.Svc.Logs(r.Context(), chi.URLParam(r, "sourceId"), r.URL.Query().Get("target"), lines)
	if search := r.URL.Query().Get("search"); search != "" {
		out = logsvc.Search(out, search)
	}
	httpx.JSON(w, http.StatusOK, map[string]any{"success": true, "logs": out})
}
